src/main/java/org/myrobotlab/codec/CodecUtils.java

```diff
@@ -20,770 +43,1252 @@
+            return String.format("org.myrobotlab.service.%s", type);
+        }
+        return type;
     }
-    return type;
-  }
 
-  public static final Set<Class<?>> WRAPPER_TYPES = new HashSet<Class<?>>(
-      Arrays.asList(Boolean.class, Character.class, Byte.class, Short.class, Integer.class, Long.class, Float.class, Double.class, Void.class));
+    /**
+     * Capitalize the first character of the given string
+     *
+     * @param line The string to be capitalized
+     * @return The capitalized version of line.
+     */
+    public static String capitalize(final String line) {
+        return Character.toUpperCase(line.charAt(0)) + line.substring(1);
+    }
 
-  public static final Set<String> WRAPPER_TYPES_CANONICAL = new HashSet<String>(
-      Arrays.asList(Boolean.class.getCanonicalName(), Character.class.getCanonicalName(), Byte.class.getCanonicalName(), Short.class.getCanonicalName(),
-          Integer.class.getCanonicalName(), Long.class.getCanonicalName(), Float.class.getCanonicalName(), Double.class.getCanonicalName(), Void.class.getCanonicalName()));
+    /**
+     * Deserializes a JSON string into the target object
+     * (or subclass of if {@link #CLASS_META_KEY} exists)
+     * using the selected JSON backend.
+     *
+     * @param json  The JSON to be deserialized in String form
+     * @param clazz The target class.
+     * @param <T>   The type of the target class.
+     * @return An object of the specified class (or a subclass of) with the state
+     * given by the json.
+     * @throws JsonDeserializationException if an error during deserialization occurs.
+     * @see #USING_GSON
+     */
+    public static <T extends Object> T fromJson(String json, Class<T> clazz) {
+        try {
+            if (USING_GSON)
+                return gson.fromJson(json, clazz);
+
+            return mapper.readValue(json, clazz);
+        } catch (Exception e) {
+            throw new JsonDeserializationException(e);
+        }
+    }
 
-  @Deprecated /* use MethodCache */
-  final static HashMap<String, Method> methodCache = new HashMap<String, Method>();
+    /**
+     * Deserializes a JSON string into the target object
+     * (or subclass of if {@link #CLASS_META_KEY} exists)
+     * using the selected JSON backend.
+     *
+     * @param json          The JSON to be deserialized in String form
+     * @param genericClass  The target class.
+     * @param parameterized The list of types used as the genericClass type parameters
+     *                      of genericClass.
+     * @param <T>           The type of the target class.
+     * @return An object of the specified class (or a subclass of) with the state
+     * given by the json.
+     * @throws JsonDeserializationException if an error during deserialization occurs.
+     * @see #USING_GSON
+     */
+    public static <T extends Object> T fromJson(String json, Class<?> genericClass, Class<?>... parameterized) {
+        try {
+            if (USING_GSON)

```

Comment:
stylistically @kwatters and I never do conditionals without curly braced blocks.  I think this adds clarity to someone say that came from a Python background and assuming indention determines blocks.
Just want to call it out, because I know an argument can be made saying it reduced eye clutter.

Reply:
Noted, I tried to find all instances of that style but I use it a lot so it's probably snuck in in a lot of other places